rust/src/bs_safe.rs

```diff
@@ -0,0 +1,535 @@
+    #[error("Could not load target library: {0}")]
+    InvalidLibrary(#[from] libloading::Error),
+    #[error("Bad Stan library version: Got {0} but expected {1}")]
+    BadLibraryVersion(String, String),
+    #[error("The Stan library was compiled without threading support. Config was {0}")]
+    StanThreads(String),
+    #[error("Failed to decode string to UTF8")]
+    InvalidString(#[from] Utf8Error),
+    #[error("Failed to construct model: {0}")]
+    ConstructFailed(String),
+    #[error("Failed during evaluation: {0}")]
+    EvaluationFailed(String),
+}
+
+type Result<T> = std::result::Result<T, BridgeStanError>;
+
+/// Open a compiled stan library.
+///
+/// The library should have been compiled with bridgestan,
+/// with the same version as the rust library.
+pub fn open_library<P: AsRef<OsStr>>(path: P) -> Result<StanLibrary> {
+    let library = unsafe { libloading::Library::new(path) }?;
+    let major: libloading::Symbol<*const c_int> = unsafe { library.get(b"bs_major_version") }?;
+    let major = unsafe { **major };
+    let minor: libloading::Symbol<*const c_int> = unsafe { library.get(b"bs_minor_version") }?;
+    let minor = unsafe { **minor };
+    let patch: libloading::Symbol<*const c_int> = unsafe { library.get(b"bs_patch_version") }?;
+    let patch = unsafe { **patch };
+    let self_major: c_int = env!("CARGO_PKG_VERSION_MAJOR").parse().unwrap();
+    let self_minor: c_int = env!("CARGO_PKG_VERSION_MINOR").parse().unwrap();
+    let self_patch: c_int = env!("CARGO_PKG_VERSION_PATCH").parse().unwrap();
+
+    if (self_major != major) | (self_minor != minor) {
+        return Err(BridgeStanError::BadLibraryVersion(
+            format!("{}.{}.{}", major, minor, patch),
+            format!("{}.{}.{}", self_major, self_minor, self_patch),
+        ));
+    }
+
+    let lib = unsafe { ffi::Bridgestan::from_library(library) }?;
+    let lib = ManuallyDrop::new(lib);
+    Ok(StanLibrary(lib))
+}
+
+/// A Stan model instance with data
+pub struct Model<T: Borrow<StanLibrary>> {
+    model: NonNull<ffi::bs_model>,
+    lib: T,
+}
+
+// Stan model is thread safe
+unsafe impl<T: Sync + Borrow<StanLibrary>> Sync for Model<T> {}
+unsafe impl<T: Send + Borrow<StanLibrary>> Send for Model<T> {}
+
+/// A random number generator for Stan
+pub struct Rng<T: Borrow<StanLibrary>> {
+    rng: NonNull<ffi::bs_rng>,
+    lib: T,
+}
+
+unsafe impl<T: Sync + Borrow<StanLibrary>> Sync for Rng<T> {}
+unsafe impl<T: Send + Borrow<StanLibrary>> Send for Rng<T> {}
+
+impl<T: Borrow<StanLibrary>> Drop for Rng<T> {
+    fn drop(&mut self) {
+        unsafe {
+            // We don't handle error messages during deconstruct
+            self.lib.borrow().0.bs_rng_destruct(self.rng.as_ptr());
+        }
+    }
+}
+
+impl<T: Borrow<StanLibrary>> Rng<T> {
+    pub fn new(lib: T, seed: u32) -> Result<Self> {
+        let mut err = ErrorMsg::new(lib.borrow());
+        let rng = unsafe {
+            lib.borrow()
+                .0
```

Comment:
I'm arbitrarily picking on this example to point out, but things like this are why I often feel weird about the tuple-style structs. I'd prefer if we used a style which had names for the members rather than just indices, though I realize it is sort of pointless for "newtype" style structs like this that only have one member.

I also can't think of a better name for the member in this case than `lib`, which means we'd have code like `lib.borrow().lib`...

Anyway, just highlighting it for discussion. I'm not sure if this is something Rust people debate or not.

Reply:
Yeah, I don't really like how that looks either. We could add a `Deref` impl so that we can drop it, but that would be visible outside the crate, and I don't think we should expose the `ffi` code in any way.

Reply:
I added a private function `ffi_lib`, which makes this a bit nicer I think.